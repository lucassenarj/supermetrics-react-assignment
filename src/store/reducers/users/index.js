import {
  USERS_GET_ARTICLES_ERROR,
  USERS_GET_ARTICLES_REQUEST,
  USERS_GET_ARTICLES_SUCCESS
} from "./../../actions";

const initialState = {
  users: [],
  loading: true,
  error: false,
};

const users = (state = initialState, action) => {
  switch(action.type) {
    case USERS_GET_ARTICLES_REQUEST: {
      return {
        ...state,
        loading: true,
        error: false,
      }
    }
    case USERS_GET_ARTICLES_ERROR: {
      return {
        ...state,
        loading: false,
        error: true,
      }
    }
    case USERS_GET_ARTICLES_SUCCESS: {
      const users = action.payload;
      return {
        ...state,
        loading: false,
        error: false,
        users,
      }
    }
    default: 
      return state;
  }
}

export default users;
