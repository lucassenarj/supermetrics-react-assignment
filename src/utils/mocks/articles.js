export const articles = [
  {
    id: 'post60cf914a226ef_28c1e277',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'try disability admiration patrol menu default shy civilian full borrow ecstasy race notice popular oak treaty',
    type: 'status',
    created_time: '2021-06-14T04:33:13+00:00'
  },
  {
    id: 'post60cf914a227b8_e716f0e6',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'carbon faithful cigarette location contrast loud full proud confine sample building prevent calculation team pedestrian swell church relevance margin ministry evolution drill climate transport advice preparation corruption expertise treasurer extension handy electron pray brake extinct wreck mystery timber convince psychology herd area sweet ban access mine seed penny crew faithful tap publisher stubborn',
    type: 'status',
    created_time: '2021-06-08T16:27:02+00:00'
  },
  {
    id: 'post60cf914a22889_039dd6d8',
    from_name: 'Carly Alvarez',
    from_id: 'user_6',
    message: 'network fuel prestige hotdog terms line agile rhythm vague point lion tap toss size bend lion breast indoor underline clock tendency bay value tin fit realize crusade revolution dog racism vertical beer shorts initiative loose language reliance mile neighborhood roar stain rear fool deficiency rubbish fine ecstasy axis orchestra draw morning wine midnight lose restrict chaos sustain south recording train brick eyebrow say film franchise gift lamb drop title treasurer intermediate ton prisoner sailor rhythm printer',
    type: 'status',
    created_time: '2021-06-01T22:15:33+00:00'
  }
];
