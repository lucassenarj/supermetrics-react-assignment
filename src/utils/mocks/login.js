export const loginResponse = {
  email: "test@test.com",
  sl_token: "smslt_f893343e131e9_167c07f380c7305",
  expiration: "1624369506541",
  name: "User test",
};
