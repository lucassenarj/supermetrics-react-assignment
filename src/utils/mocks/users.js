export const users = [
  {
    user_id: 'user_4',
    user_name: 'Britany Heise',
    posts: [
      {
        id: 'post60cf914a225f1_37d9a0b4',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'conflict fight bend resident term fly coincidence put wagon west recruit faithful term rank coincidence guideline west coincide nuance roar race treasurer freedom cottage flow boat tired term definition smell forget facility final chest lip belong clearance printer qualified loose satellite say depression difficult point sailor dismissal leak safety',
        type: 'status',
        created_time: '2021-06-20T15:39:53+00:00'
      },
      {
        id: 'post60cf914a2262f_62849f6b',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'tick genuine racism leader prize retired suitcase dominant syndrome retired speed flawed beer book corn rehearsal critic union hiccup allocation variable belong beg snub midnight scheme accumulation penny tap',
        type: 'status',
        created_time: '2021-06-19T18:05:56+00:00'
      },
      {
        id: 'post60cf914a2269d_4856832f',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'smash suntan delicate irony cave superintendent speed satellite opposition fit thoughtful galaxy duck braid seed meadow habitat cheese product beer television giant competition barrier cluster trace mine bolt bike album loud spell resident write delay appear egg white patrol hand prisoner decorative desert braid anger pray deficiency tap',
        type: 'status',
        created_time: '2021-06-16T05:46:19+00:00'
      },
      {
        id: 'post60cf914a226e0_20f5688b',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'sword absorb awful bracket urine gradient pour plagiarize uncle adoption core makeup',
        type: 'status',
        created_time: '2021-06-14T14:10:05+00:00'
      },
      {
        id: 'post60cf914a226e5_1f523234',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'convention toss glory convince candle blade throne album reconcile terrify mess reliable multimedia smash linen agriculture bullet tin fool brave safety egg white introduce jurisdiction story jurisdiction velvet bend epicalyx format refuse superintendent solo boy ballot prescription abstract heal palm rescue safety mirror combine arch waist tie dressing sanctuary porter graze spend canvas printer computing scholar television second harmful',
        type: 'status',
        created_time: '2021-06-14T08:58:40+00:00'
      },
      {
        id: 'post60cf914a22747_d34228f9',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'noble dead index palm transmission aluminium pardon flourish herb wreck fly yard confront revoke introduce information speech spend jealous bolt rank cave relation swipe scream stain evening prisoner chocolate pedestrian crossing energy college upset gain conservative absent mystery slab menu drop',
        type: 'status',
        created_time: '2021-06-12T05:45:08+00:00'
      },
      {
        id: 'post60cf914a2277e_40aa13fb',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'good establish fill deadly executive',
        type: 'status',
        created_time: '2021-06-11T04:48:34+00:00'
      },
      {
        id: 'post60cf914a22816_1a4b4e6a',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'generation race talkative mile discourage dignity survey home mastermind conception debt bolt cluster bell water torture pool popular boy host vegetarian bother abstract broadcast tap palm glare blue jean horseshoe hate age final reinforce sailor cage kill parachute album convention field album trail federation tired survivor buy flower innocent environmental manage prize viable underline animal tile charm brick freckle psychology bar entitlement',
        type: 'status',
        created_time: '2021-06-05T17:59:55+00:00'
      },
      {
        id: 'post60cf914a2283f_43e890b1',
        from_name: 'Britany Heise',
        from_id: 'user_4',
        message: 'pick axis cause achievement rule fame hotdog multimedia voyage line sacrifice transport dream delay patrol leave impulse dominant fame voyage sight format bill grip trouble lake run wisecrack crop addition tract tense entitlement chord loud ecstasy diamond handy district network flow carry officer baby franchise dressing peasant popular variant interest achievement benefit transport charm carve pick superintendent breed squash sensation bishop debate crusade rubbish evening',
        type: 'status',
        created_time: '2021-06-04T12:39:09+00:00'
      }
    ]
  },
  {
    user_id: 'user_6',
    user_name: 'Carly Alvarez',
    posts: [
      {
        id: 'post60cf914a226ef_28c1e277',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'try disability admiration patrol menu default shy civilian full borrow ecstasy race notice popular oak treaty',
        type: 'status',
        created_time: '2021-06-14T04:33:13+00:00'
      },
      {
        id: 'post60cf914a227b8_e716f0e6',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'carbon faithful cigarette location contrast loud full proud confine sample building prevent calculation team pedestrian swell church relevance margin ministry evolution drill climate transport advice preparation corruption expertise treasurer extension handy electron pray brake extinct wreck mystery timber convince psychology herd area sweet ban access mine seed penny crew faithful tap publisher stubborn',
        type: 'status',
        created_time: '2021-06-08T16:27:02+00:00'
      },
      {
        id: 'post60cf914a22889_039dd6d8',
        from_name: 'Carly Alvarez',
        from_id: 'user_6',
        message: 'network fuel prestige hotdog terms line agile rhythm vague point lion tap toss size bend lion breast indoor underline clock tendency bay value tin fit realize crusade revolution dog racism vertical beer shorts initiative loose language reliance mile neighborhood roar stain rear fool deficiency rubbish fine ecstasy axis orchestra draw morning wine midnight lose restrict chaos sustain south recording train brick eyebrow say film franchise gift lamb drop title treasurer intermediate ton prisoner sailor rhythm printer',
        type: 'status',
        created_time: '2021-06-01T22:15:33+00:00'
      }
    ]
  },
];