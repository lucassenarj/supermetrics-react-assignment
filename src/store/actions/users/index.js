import api from "./../../../services/api";
import { groupPostByUsers } from "./../../../utils/helpers";
import { setArticleList } from "./../articles";

export const USERS_GET_ARTICLES_ERROR = "USERS_GET_ARTICLES_ERROR";
export const USERS_GET_ARTICLES_REQUEST = "USERS_GET_ARTICLES_REQUEST";
export const USERS_GET_ARTICLES_SUCCESS = "USERS_GET_ARTICLES_SUCCESS";

export function getArticles(sl_token) {
  return async (dispatch) => {
    try {
      dispatch({
        type: USERS_GET_ARTICLES_REQUEST,
      });

      const response = await api.get(`/posts?sl_token=${sl_token}`);

      if (response.status === 200) {
        const data = groupPostByUsers(response.data.data.posts);

        dispatch(setArticleList(data[0].posts));
  
        dispatch({
          type: USERS_GET_ARTICLES_SUCCESS,
          payload: data,
        });
      }

    } catch (error) {
      dispatch({
        type: USERS_GET_ARTICLES_ERROR,
        payload: error,
      });
    }
  }
}