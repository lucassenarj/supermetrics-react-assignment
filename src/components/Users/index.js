import React from "react";
import User from "../User";
import { useDispatch, useSelector } from "react-redux";
import { setArticleList } from "./../../store/actions";
import "./index.scss";

function Users() {
  const dispatch = useDispatch();
  const { users } = useSelector(state => state.users);
  return (
    <div>
      { users.map((user) => (
        <User
          name={user.user_name}
          posts={user.posts.length}
          key={user.user_id}
          id={user.user_id}
          onClick={() => dispatch(setArticleList(user.posts))}
        />
      ))}
    </div>
  );
};

export default Users;
