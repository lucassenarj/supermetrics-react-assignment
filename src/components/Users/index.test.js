import React from "react";
import configureStore from "redux-mock-store";
import { render, fireEvent } from "@testing-library/react";
import * as redux from "react-redux";
import { users } from "./../../utils/mocks/users";
import { MemoryRouter } from "react-router";
import Users from "./";

const mockStore = configureStore([]);

describe("Components", () => {
  describe("Users", () => {
    let wrapper;
    let store;
    const dispatch = jest.fn();
  
    beforeEach(() => {
      store = mockStore({
        users: {
          users
        }
      });

      jest.spyOn(redux, "useDispatch").mockImplementation(() => dispatch);

      wrapper = render(
        <redux.Provider store={store}>
          <MemoryRouter>
            <Users users={users} />
          </MemoryRouter>
        </redux.Provider>
      );
    });
    it("renders correctly", () => {
      const { container } = wrapper;
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should click button User button", () => {
      const { user_id } = users[0];
      const { getByTestId } = wrapper;
      fireEvent.click(getByTestId(user_id));

      expect(redux.useDispatch).toHaveBeenCalled();
      expect(redux.useDispatch).toHaveBeenCalledTimes(1);
    });
  });
});
