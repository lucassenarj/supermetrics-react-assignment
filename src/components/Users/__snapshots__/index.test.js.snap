// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Users renders correctly 1`] = `
<div>
  <button
    class="user"
    data-testid="user_4"
  >
    <p>
      Britany Heise
       
    </p>
    <p
      class="user__count"
    >
      9
    </p>
  </button>
  <button
    class="user"
    data-testid="user_6"
  >
    <p>
      Carly Alvarez
       
    </p>
    <p
      class="user__count"
    >
      3
    </p>
  </button>
</div>
`;
